enhancements/20240523-support-rwx-volume-to-downstream-cluster.md

```diff
@@ -0,0 +1,109 @@
+
+### API changes
+TBD
+
+## Design
+
+### Implementation Overview
+
+A quick overview of the design could be checked below:
+
+![Downstream RWX Support](20240523-support-rwx-volume-to-downstream-cluster/general-rwx.png)
+
+### Solution 1.
+
+Like the RWO solution, we would like to leverage the libvirt capability to hotplug the filesystem and then use the workload pod as a host mountpoint, but currently, it is not supported. The hotplug filesystem on the kubevirt side is referred to as virtiofs. However, virtiofs does not support live migration, so we could not use this solution.
+
+Related feature request to Kubevirt: https://github.com/kubevirt/kubevirt/issues/11699
+
+### Solution 2.
+
+The second solution is like the Longhorn CSI behavior with RWX volume. The Longhorn share manager will provide the NFS endpoint through ClusterIP. The thing we need to take care of is network space. The downstream cluster uses a different network space from the Harvester/Longhorn cluster. We need to ensure the downstream cluster can interact with the NFS endpoint with the same network space. Like below:
+
+![Network Space](20240523-support-rwx-volume-to-downstream-cluster/rwx-network.png)
+
+The detailed flowchart is like below:
+
+- RWX volume Create:
+
+![RWX volume Create](20240523-support-rwx-volume-to-downstream-cluster/create-rwx-volume.png)
+
+- RWX volume Attachment:
+
+![RWX volume Attachment](20240523-support-rwx-volume-to-downstream-cluster/attach-rwx-volume.png)
+
+### Each component detailed design
+
+#### 1. Longhorn
+Longhorn share-manager can provide the NFS endpoint through the storage network.
+
+Related issue: https://github.com/longhorn/longhorn/issues/8184
+
+#### 2. Harvester CSI Driver
+Harvester CSI Driver need to add capability to create the RWX volume through Longhorn client. Also, we should add the corresponding handles on NodeStage/NodePublish stage. The NodeStage will mount the NFS endpoint to the host as global path. Then, the NodePublish will bind mount the global path for the workload.
+
+The whole flow might be like that when get the request to create RWX volume, Harvester CSI Driver need to create corresponding RWX volume on Longhorn. Then, the Harvester CSI driver will create a global path on NodeStage. Last, with NodePublish, the Harvester CSI driver will bind mount the global path for the workload.
+
+#### 3. NetworkFS controller on Harvester (New Added)
+We need to create a workload pod to bring up the Longhorn-share manager so that the Longhorn-share-manager can create the NFS endpoint.
+We would like to introduce the NetworkFS controller to handle(create) the pod. The advantage with this component is that we could control the whole network filesystem on Harvester side. The network filesystem could be NFS. Later, we could add more network filesystem like CIFS(SMB)/CephFS. Also, we can add some network strategies like network policy or service endpoint to limit the access to the network filesystem.
```

Comment:
Can you explain more about how to expose the LH NFS endpoint to the downstream cluster during CSI `NodeStage`? Does it require the LH feature [feat(LEP): storage network for RWX volumes](https://github.com/longhorn/longhorn/pull/8616)?

Thanks

Reply:
Update the networkfs-manager and the responsibility of its CRD for the endpoint information.